### PR TITLE
NemSDK

### PR DESCRIPTION
**Pull request by RobertoSnap**

### CODE REVIEW

File: config/nem.php [hunk 1]
```diff
@@ -72,14 +72,16 @@
...
             "use_ssl" => false,
             "host" => "127.0.0.1",
             "port" => 7890,
             "endpoint" => "/",
         ],
         'testing' => [
+        	"protocol" => "http",
             "use_ssl" => false,
-            "host" => "go.nem.ninja",
+            "host" => "10.0.2.2",
```
Comment (evias) [inline config/nem.php]: is it possible to revert this change? Not sure but I guess the 10.X IP is a local node of yours.

File: src/Infrastructure/Network.php [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+		$this->checkNetwork($identifier);
+
+		return $this->networkType;
+	}
+
+	private function checkNetwork( $identifier = null ) {
+		if ( is_string( $identifier ) ) {
+			$identifier = strtoupper( $identifier );
+			switch ( true ) {
+				case strlen( $identifier ) <= 46 && strlen( $identifier ) >= 40:
```
Comment (evias) [inline src/Infrastructure/Network.php]: would be better to use regular expressions here

File: src/Infrastructure/Network.php [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+namespace evias\NEMBlockchain\Infrastructure;
+
+use evias\NEMBlockchain\NemSDK;
+
+class Network {
+
+	public $nemSDK;
+	public $version;
+	public $version2;
```
Comment (evias) [inline src/Infrastructure/Network.php]: instead of having version and version2, i suggest using an array of *versions IDs per network*

File: src/Infrastructure/TransactionMultisig.php [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+namespace evias\NEMBlockchain\Infrastructure;
+
+use evias\NEMBlockchain\NemSDK;
+use evias\NEMBlockchain\Models\Transaction\Multisig;
+use evias\NEMBlockchain\Models\Transaction\Transfer;
+use evias\NEMBlockchain\Models\Transaction\Mosaic;
+
+
+
+class TransactionMultisig {
```
Comment (evias) [inline src/Infrastructure/TransactionMultisig.php]: why the need for a specific class for Multisig transactions class and another Multisig class? I believe it is best to keep it all in the Transaction class and let the transaction class handle wrapping like this class here handles it.

In Transaction::multisig() you would wrap the transaction in a "otherTrans" attribute. I think handling multisigs as *specific objects* is wrong as it will also split fees definition etc.. about transactions in several classes making it too complicated to work with.

File: src/Infrastructure/Network.php (hunk 2 above)
Comment (RobertoSnap) [inline src/Infrastructure/Network.php]:
```
	private function checkNetwork( $identifier = null ) {
		if(preg_match('/^([A-Z]{1})[A-Z0-9-]{39,45}$/', strtoupper($identifier),$matches)){
			if($matches[1] === "N") {
				$this->setNetworkType( "MAINNET" );
			} elseif ($matches[1] === "T") {
				$this->setNetworkType( "TESTNET" );
			} else {
				throw new \Exception( "Could not indentify Nem Network type from identifier. Assumed string and address." );
			}
		} else {
			$networkId = ($this->nemSDK->node()->info())->metaData->networkId;
			if ( $networkId === 104 ) {
				$this->setNetworkType( "main" );
			} elseif ( $networkId === - 104 ) {
				$this->setNetworkType( "test" );
			} else {
				throw new \Exception( "Could not indentify Nem Network type from calling the node. Assumed nothing as identfier" );
			}
		}
	}
```

File: src/Infrastructure/Network.php (hunk 3 above)
Comment (RobertoSnap) [inline src/Infrastructure/Network.php]:
```php
public function getVersion( $identifier = null , $version = 1) {
		if(!empty($this->version[$version])){
			return $this->version[$version];
		}
		$this->checkNetwork($identifier);

		return $this->version[$version];
	}

```

File: config/nem.php (hunk 1 above)
Comment (RobertoSnap) [inline config/nem.php]: Fixed

File: src/Infrastructure/TransactionMultisig.php (hunk 4 above)
Comment (RobertoSnap) [inline src/Infrastructure/TransactionMultisig.php]:
```php
class Transaction {
	public $nemSDK;
	private $multisig = false;
	private $transaction;

	public function __construct( NemSDK $nemSDK ) {
		$this->nemSDK = $nemSDK;
	}

	/*Multisig*/
	public function multisig( $multisigPublicKey, $multisigPrivateKey ) {
		$this->transaction = new Multisig($this->nemSDK, $multisigPublicKey, $multisigPrivateKey );
		$this->multisig = true;
		return $this;
	}


	public function transfer( $recipient, $amount, $message, $senderPublicKey, $senderPrivateKey = null, $mosaics = array(), $fee = "auto" , $encrypted = false) {
		if($this->multisig){
			$this->transaction->setOtherTrans((new Transfer($this->nemSDK, $recipient, $amount, $message, $senderPublicKey, $this->transaction->multisigPrivateKey, $fee, $encrypted, $mosaics))->toDTO());
		}else {
			$this->transaction = new Transfer($this->nemSDK, $recipient, $amount, $message, $senderPublicKey, $senderPrivateKey, $fee, $encrypted, $mosaics);
		}
		return $this->sendTransaction();
	}

	public function mosaic( $mosaicName, $mosaicDescription, $namespaceId, $signerPublicKey, $signerPrivateKey = null, $properties = array(), $levy = array() ) {
		if($this->multisig){
			$this->transaction->setOtherTrans((new Mosaic($this->nemSDK, $mosaicName, $mosaicDescription, $namespaceId, $signerPublicKey, $this->transaction->multisigPrivateKey, $properties, $levy ))->toDTO());
		}else {
			$this->transaction = new Mosaic($this->nemSDK, $mosaicName, $mosaicDescription, $namespaceId, $signerPublicKey, $signerPrivateKey, $properties, $levy );
		}
		return $this->sendTransaction();
	}

	private function sendTransaction(){
		return json_decode($this->nemSDK->api->post('/transaction/prepare-announce',$this->transaction->toDTO() ));
	}

}
```

**Comment by evias** [comment]
squashed in https://github.com/evias/php-nem-laravel/pull/2
